File: app/src/main/java/com/dingaimin/dedaocompanion/QueuePolicy.java
package com.dingaimin.dedaocompanion;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

final class QueuePolicy {
    static List<RedPacketItem> apply(List<RedPacketItem> source, String course,
                                     boolean hideCompleted, boolean newestFirst) {
        ArrayList<RedPacketItem> result = new ArrayList<>();
        if (source == null) return result;
        String selectedCourse = course == null || course.isBlank() ? "全部课程" : course;
        for (RedPacketItem item : source) {
            if (!"全部课程".equals(selectedCourse) && !selectedCourse.equals(item.course)) continue;
            if (hideCompleted && item.completed) continue;
            result.add(item);
        }
        Comparator<RedPacketItem> order = Comparator.comparingLong(item -> item.claimedAtSeconds);
        if (newestFirst) order = order.reversed();
        Collections.sort(result, order);
        return result;
    }

    static List<RedPacketItem> applyCustomOrder(List<RedPacketItem> source,
                                                 List<String> orderedKeys) {
        ArrayList<RedPacketItem> result = new ArrayList<>();
        if (source == null) return result;
        result.addAll(source);
        if (orderedKeys == null || orderedKeys.isEmpty()) return result;

        Map<String, Integer> ranks = new HashMap<>();
        for (int i = 0; i < orderedKeys.size(); i++) {
            String key = orderedKeys.get(i);
            if (key != null && !key.isBlank()) ranks.putIfAbsent(key, i);
        }
        result.sort(Comparator.comparingInt(
                item -> ranks.getOrDefault(stableKey(item), Integer.MAX_VALUE)));
        return result;
    }

    static String stableKey(RedPacketItem item) {
        if (item == null) return "";
        if (item.id != null && !item.id.isBlank()) return "id:" + item.id + ":" + item.type;
        return "title:" + item.course + ":" + item.title;
    }

    static List<RedPacketItem> excluding(List<RedPacketItem> source, Set<String> removedKeys) {
        ArrayList<RedPacketItem> result = new ArrayList<>();
        for (RedPacketItem item : source) {
            if (!removedKeys.contains(stableKey(item))) result.add(item);
        }
        return result;
    }

    /** Preserve the current identity; deleting it selects its successor, then its predecessor. */
    static int indexAfterRemoval(int currentIndex, int removedIndex, int originalSize) {
        if (originalSize <= 1) return -1;
        int adjusted = currentIndex > removedIndex ? currentIndex - 1 : currentIndex;
        return Math.max(0, Math.min(adjusted, originalSize - 2));
    }

    static List<String> mergeVisibleOrder(List<String> globalOrder,
                                          List<String> reorderedVisibleKeys) {
        ArrayList<String> result = new ArrayList<>();
        if (globalOrder != null) result.addAll(globalOrder);
        if (reorderedVisibleKeys == null || reorderedVisibleKeys.isEmpty()) return result;
        Set<String> visible = new LinkedHashSet<>(reorderedVisibleKeys);
        int replacement = 0;
        for (int i = 0; i < result.size() && replacement < reorderedVisibleKeys.size(); i++) {
            if (visible.contains(result.get(i))) {
                result.set(i, reorderedVisibleKeys.get(replacement++));
            }
        }
        return result;
    }

    private QueuePolicy() {
    }
}

File: app/src/main/java/com/dingaimin/dedaocompanion/QueuePolicy.kt
package com.dingaimin.dedaocompanion

/** Pure queue operations: never mutate the source cache or the official player's library. */
object QueuePolicy {
    fun apply(
        source: List<RedPacketItem>?,
        course: String?,
        hideCompleted: Boolean,
        newestFirst: Boolean,
    ): List<RedPacketItem> {
        val selectedCourse = course?.takeUnless { it.isBlank() } ?: "全部课程"
        val filtered =
            source.orEmpty().filter {
                (selectedCourse == "全部课程" || selectedCourse == it.course) &&
                    (!hideCompleted || !it.completed)
            }
        return if (newestFirst) filtered.sortedByDescending { it.claimedAtSeconds }
        else filtered.sortedBy { it.claimedAtSeconds }
    }

    fun applyCustomOrder(
        source: List<RedPacketItem>?,
        orderedKeys: List<String>?,
    ): List<RedPacketItem> {
        val ranks = linkedMapOf<String, Int>()
        orderedKeys.orEmpty().forEachIndexed { index, key ->
            if (key.isNotBlank()) ranks.putIfAbsent(key, index)
        }
        return source.orEmpty().sortedBy { ranks[stableKey(it)] ?: Int.MAX_VALUE }
    }

    fun stableKey(item: RedPacketItem?): String =
        when {
            item == null -> ""
            item.id.isNotBlank() -> "id:${item.id}:${item.type}"
            else -> "title:${item.course}:${item.title}"
        }

    fun excluding(source: List<RedPacketItem>, removedKeys: Set<String>): List<RedPacketItem> =
        source.filterNot { stableKey(it) in removedKeys }

    /** Preserve the current identity; deleting it selects its successor, then its predecessor. */
    fun indexAfterRemoval(currentIndex: Int, removedIndex: Int, originalSize: Int): Int {
        if (originalSize <= 1) return -1
        val adjusted = if (currentIndex > removedIndex) currentIndex - 1 else currentIndex
        return adjusted.coerceIn(0, originalSize - 2)
    }

    fun mergeVisibleOrder(
        globalOrder: List<String>?,
        reorderedVisibleKeys: List<String>?,
    ): List<String> {
        val result = globalOrder.orEmpty().toMutableList()
        val reordered = reorderedVisibleKeys.orEmpty()
        val visible = reordered.toSet()
        var replacement = 0
        result.indices.forEach { index ->
            if (result[index] in visible && replacement < reordered.size) {
                result[index] = reordered[replacement++]
            }
        }
        return result
    }
}
